### PR TITLE
Siyun/string description

### PR DESCRIPTION
**Pull request by Siyun96**
Implementation for issue #24. 

For each code, we look up its text description in MRCONSO.RRF (STR column). Since there might be multiple entries for the same code, we only keep the shortest string (with at least 4 characters) that satisfy the query condition: `"SUPPRESS" NOT IN ('O', 'Y') AND ISPREF='Y' AND LAT='ENG'`. For code with no description, they are mapped to `NO_DEF`. The mapping is stored as an `OntologyCodeDictionary` in `ontology.db`.

`create_timelines.py` shows an example of how to query per-patient timeline object with text descriptions. 

`test_patient.json` shows an example output from a SynPUF patient.

### CODE REVIEW

File: ehr_ml/create_timelines.py
Comment (Lalaland) [inline ehr_ml/create_timelines.py:24]: This should be updated?

File: ehr_ml/create_timelines.py
Comment (Lalaland) [inline ehr_ml/create_timelines.py:40]: Please provide the second argument to TimelineReader by setting readall=False

File: ehr_ml/create_timelines.py
Comment (Lalaland) [inline ehr_ml/create_timelines.py:55]: Printing seems a bit awkward here for this sort of tool. I would remove this.

File: ehr_ml/create_timelines.py
Comment (Lalaland) [inline ehr_ml/create_timelines.py:70]: Please add the patient id to this json structure.

File: ehr_ml/create_timelines.py
Comment (Lalaland) [inline ehr_ml/create_timelines.py:72]: This is a bad structure. Please use a list here instead of a custom key for each day.

patient_timeline['days'] = []
for day in patient.days:
    patient_timeline['days'].append({ ...}

File: ehr_ml/create_timelines.py
Comment (Lalaland) [inline ehr_ml/create_timelines.py:73]: Use date.isoformat() instead of str() here.

File: ehr_ml/create_timelines.py
Comment (Lalaland) [inline ehr_ml/create_timelines.py:79]: Why aren't you including the ontology text description stuff here as well?

File: native/reader.h
Comment (Lalaland) [inline native/reader.h:219]: Why do we need a new class for this?

File: pyproject.toml
Comment (Lalaland) [inline pyproject.toml:18]: Please rename this to "convert_patient_to_json" or something.

**Review by Lalaland — COMMENTED** [review]
Nice work!

I think this is the right idea, but there are a couple of things that should be fixed before merging. Most important would be fixing the json format you are generating. "day 0", "day 1", etc is going to cause a lot of problems on the consumption side.

File: native/reader.h
Comment (Siyun96) [reply native/reader.h:219]: TermDictionary maps from string to uint32. For code descriptions, I need string to string mapping.

File: ehr_ml/create_timelines.py
Comment (Siyun96) [reply ehr_ml/create_timelines.py:79]: I'm only testing on SynPUF. It seems the "observation with values" have code "Visit/OP" and the text description returns "NO_DEF". Do we expect there to be a description for that? I thought this might be some placeholder we add (similar to Race and Gender). Nonetheless, I added the code so that "observation with values" have the same format as "observation".

File: pyproject.toml
Comment (Siyun96) [reply pyproject.toml]: Changed to "inspect_patient_timeline".

File: ehr_ml/create_timelines.py
Comment (Lalaland) [reply ehr_ml/create_timelines.py]: There wouldn't be a text description for that, but lab values will tend to have text descriptions (stuff like blood pressure tests, etc).

File: native/reader.h
Comment (Lalaland) [reply native/reader.h:219]: You should make this a C++ template so that you can have the same code for string and uint32.

File: pyproject.toml
Comment (Lalaland) [reply pyproject.toml]: Inspect patient timeline? Wouldn't that be confusing with the existing inspect_timeline?

**Review by Lalaland — COMMENTED** [review]
Mostly good. Just fix the two final things (the name of the program and use a C++ template to avoid code duplication).

**Review by jason-fries — APPROVED** [review]
LGTM! As long as @Lalaland has signed off here I think it's fine to merge.

**Review by Lalaland — COMMENTED** [review]
Looks good to me! Feel free to merge when you are ready.

**Pull request merged**
